File: tree/BinarySearchTree(BST)/BST.java
class Node<T> {
  public T data;
  public Node<T> left;
  public Node<T> right;

  public Node(T value) {
      data = value;
      left = null;
      right = null;
  }
}

class BST<T extends Comparable<T>> {
  private Node<T> root;

  private Node<T> findMin(Node<T> node) {
      if (node.left == null) {
          return node;
      }
      return findMin(node.left);
  }

  private Node<T> deleteNode(Node<T> root, T value) {
      if (root == null) {
          return root;
      }

      if (value.compareTo(root.data) < 0) {
          root.left = deleteNode(root.left, value);
      } else if (value.compareTo(root.data) > 0) {
          root.right = deleteNode(root.right, value);
      } else {
          if (root.left == null) {
              return root.right;
          } else if (root.right == null) {
              return root.left;
          }

          Node<T> temp = findMin(root.right);
          root.data = temp.data;
          root.right = deleteNode(root.right, temp.data);
      }
      return root;
  }

  public void deleteMethod(T value) {
      root = deleteNode(root, value);
  }

  public void insert(T value) {
      Node<T> newNode = new Node<>(value);

      if (root == null) {
          root = newNode;
      } else {
          Node<T> current = root;
          Node<T> parent = null;

          while (current != null) {
              parent = current;
              if (value.compareTo(current.data) < 0) {
                  current = current.left;
              } else {
                  current = current.right;
              }
          }

          if (value.compareTo(parent.data) < 0) {
              parent.left = newNode;
          } else {
              parent.right = newNode;
          }
      }
  }

  public boolean search(T value) {
      Node<T> current = root;
      while (current != null) {
          if (value.compareTo(current.data) == 0) {
              return true;
          } else if (value.compareTo(current.data) < 0) {
              current = current.left;
          } else {
              current = current.right;
          }
      }
      return false;
  }

  private void inOrderHelper(Node<T> node) {
      if (node != null) {
          inOrderHelper(node.left);
          System.out.print(node.data + " ");
          inOrderHelper(node.right);
      }
  }

  public void inOrder() {
      inOrderHelper(root);
      System.out.println();
  }

  private void preOrderHelper(Node<T> node) {
      if (node != null) {
          System.out.print(node.data + " ");
          preOrderHelper(node.left);
          preOrderHelper(node.right);
      }
  }

  public void preOrder() {
      preOrderHelper(root);
      System.out.println();
  }

  private void postOrderHelper(Node<T> node) {
      if (node != null) {
          postOrderHelper(node.left);
          postOrderHelper(node.right);
          System.out.print(node.data + " ");
      }
  }

  public void postOrder() {
      postOrderHelper(root);
      System.out.println();
  }
}

class Solution {
  public static void main(String[] args) {
      BST<Integer> obj = new BST<>();
      obj.insert(5);
      obj.insert(3);
      obj.insert(7);
      obj.insert(2);
      obj.insert(4);
      obj.insert(6);
      obj.insert(8);

      obj.inOrder();
      obj.preOrder();
      obj.postOrder();

      obj.deleteMethod(2);

      obj.inOrder();
  }
}

File: tree/BinarySearchTree(BST)/BST.py
class Node:
    def __init__(self, value):
        self.data = value
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None
    
    def find_min(self, node):
        if node.left is None:
            return node
        return self.find_min(node.left)
    
    def delete_node(self, root, value):
        if root is None:
            return root
        
        if value < root.data:
            root.left = self.delete_node(root.left, value)
        elif value > root.data:
            root.right = self.delete_node(root.right, value)
        else:
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            
            temp = self.find_min(root.right)
            root.data = temp.data
            root.right = self.delete_node(root.right, temp.data)
        
        return root
    
    def delete_method(self, value):
        self.root = self.delete_node(self.root, value)
    
    def insert(self, value):
        new_node = Node(value)
        
        if self.root is None:
            self.root = new_node
        else:
            current = self.root
            parent = None
            
            while current is not None:
                parent = current
                if value < current.data:
                    current = current.left
                else:
                    current = current.right
            
            if value < parent.data:
                parent.left = new_node
            else:
                parent.right = new_node
    
    def search(self, value):
        current = self.root
        while current is not None:
            if current.data == value:
                return True
            elif value < current.data:
                current = current.left
            else:
                current = current.right
        return False
    
    def in_order_helper(self, node):
        if node is not None:
            self.in_order_helper(node.left)
            print(node.data, end=" ")
            self.in_order_helper(node.right)
    
    def in_order(self):
        self.in_order_helper(self.root)
        print()
    
    def pre_order_helper(self, node):
        if node is not None:
            print(node.data, end=" ")
            self.pre_order_helper(node.left)
            self.pre_order_helper(node.right)
    
    def pre_order(self):
        self.pre_order_helper(self.root)
        print()
    
    def post_order_helper(self, node):
        if node is not None:
            self.post_order_helper(node.left)
            self.post_order_helper(node.right)
            print(node.data, end=" ")
    
    def post_order(self):
        self.post_order_helper(self.root)
        print()

# Main function
if __name__ == "__main__":
    obj = BST()
    obj.insert(5)
    obj.insert(3)
    obj.insert(7)
    obj.insert(2)
    obj.insert(4)
    obj.insert(6)
    obj.insert(8)
    
    obj.in_order()
    obj.pre_order()
    obj.post_order()

    obj.delete_method(2)

    obj.in_order()
